lib/voight_kampff/test.rb

```diff
@@ -1,7 +1,45 @@
+# frozen_string_literal: true
+
 module VoightKampff
+  # Test User-Agent against Voight-Kampff
   class Test
     CRAWLERS_FILENAME = 'crawler-user-agents.json'
 
+    class << self
+      def crawlers
+        @crawlers ||= JSON.parse File.read preferred_path
+      end
+
+      def crawler_regexp
+        @crawler_regexp ||= begin
+          # NOTE: This is admittedly a bit convoluted
+          # but the performance gains make it worthwhile
+          crawler_patterns =
+            crawlers.map.with_index do |crawler, index|
+              "(?<match#{index}>#{crawler['pattern']})"
+            end.join('|')
+          crawler_patterns = "(#{crawler_patterns})"
+          Regexp.new(crawler_patterns, Regexp::IGNORECASE)
+        end
+      end
+
+      def preferred_path
+        # These paths should be orderd by priority
+        [
+          (Rails.root if defined? Rails),
+          Dir.pwd,
+          VoightKampff::ROOT
+        ]
+          .find do |base_path|
+            next unless base_path
+
+            path = File.join base_path, 'tmp', CRAWLERS_FILENAME
+
+            break path if File.exist? path
+          end
+      end
```

Comment:
I'm not sure why the old `lookup_paths` and `preferred_path` methods were changed. If there wasn't a good reason for the change, I feel like the old methods were better broken up into logical units and were more readable as well.

Reply:
> I'm not sure why the old `lookup_paths` and `preferred_path` methods were changed.

`lookup_paths` becomes too simple.

> If there wasn't a good reason for the change, I feel like the old methods were better broken up into logical units and were more readable as well.

The previous version had `join` inside `lookup_paths`, but now paths are mostly just Strings, so the `join` operation is in `preferred_path`'s `find`, after condition. But I've returned `lookup_paths` with new content, as you wish.